### PR TITLE
[DO NOT MERGE] Survey improvements

### PR DESCRIPTION
**Pull request by elenatanasoiu**
For: https://trello.com/c/n8QbgGXK/169-make-general-improvements-to-surveys-js

## What
This PR adds various improvements that did not make into the surveys during the development of this card: https://trello.com/c/8YSfHXc2. 

## Why 
The work for the aforementioned card was ultimately not used, but some of the changes made there to surveys are still viable and have been extracted out in this PR.

### CODE REVIEW

File: spec/javascripts/surveys-spec.js [hunk 1]
```diff
@@ -692,9 +692,9 @@ describe('Surveys', function () {
         var testSurvey = {
           startTime: new Date('July 5, 2016').getTime(),
           endTime: new Date('July 10, 2016 23:50:00').getTime(),
-          activeWhen: function () { return false },
           url: 'example.com/small-survey'
         }
+        spyOn(surveys, 'activeWhen').and.returnValue(false)
```
Comment (h-lame) [inline spec/javascripts/surveys-spec.js]: Now that a function isn't allowed for `activeWhen` on surveys this is a good way to fix this test, but I think we should do the same to the sibling test above: "includes the survey when the activeWhen function returns true".  That test is passing "just because" and it'd be better to be more explicit.

File: doc/surveys.md [hunk 2]
```diff
@@ -39,89 +41,119 @@ This is used both as the Google Analytics tag to record events against, and used
...
 ```html
 <section id="user-satisfaction-survey" class="visible" aria-hidden="false">
   <div class="survey-wrapper">
     <a class="survey-close-button" href="#user-survey-cancel" aria-labelledby="survey-title user-survey-cancel" id="user-survey-cancel" role="button">Close</a>
-    <h2 class="survey-title" id="survey-title">Heading copy</h2>
+    <h2 class="survey-title" id="survey-title">{{title}}</h2>
     <p>
-      <a class="survey-primary-link" href="javascript:void()" id="take-survey" target="_blank" rel="noopener noreferrer">Link text</a>
-      This will open a short survey on another website
+      <a class="survey-primary-link" href="" id="take-survey" target="_blank" rel="noopener noreferrer">{{surveyCta}}</a>
```
Comment (h-lame) [inline doc/surveys.md]: I think we're missing the `{{surveyUrl}}` template parameter in the `href` here.

File: app/assets/javascripts/surveys.js [hunk 3]
```diff
@@ -6,51 +6,55 @@
...
-    return '<a ' + className + ' href="javascript:void()" id="take-survey" target="_blank" rel="noopener noreferrer">' + text + '</a>'
+    return '<a ' + className + ' href="{{surveyUrl}}" id="take-survey" target="_blank" rel="noopener noreferrer">' + text + '</a>'
   }
 
   var templateBase = function (children) {
     return (
       '<section id="user-satisfaction-survey" class="visible" aria-hidden="false">' +
       '  <div class="survey-wrapper">' +
       '    <a class="survey-close-button" href="#user-survey-cancel" aria-labelledby="survey-title user-survey-cancel" id="user-survey-cancel" role="button">Close</a>' +
-      '    <h2 class="survey-title" id="survey-title">Tell us what you think of GOV.UK</h2>' +
```
Comment (h-lame) [inline app/assets/javascripts/surveys.js]: We can leave the title in here can't we?  Or do we only apply template params before we pass things into `templateBase`?

File: app/assets/javascripts/surveys.js [hunk 4]
```diff
@@ -189,6 +193,65 @@
...
+      return {
+        title: "Tell us what you think of GOV.UK",
+        surveyCta: "Take the 3 minute survey",
+        surveyCtaPostscript: "This will open a short survey on another website",
+        render: function(survey) {
+          var templateArgs = (survey.templateArguments || {}),
+              surveyUrl = userSurveys.addCurrentPathtoURL(survey);
+
+          // Smart survey can record the URL of the survey link if passed through as a query param
+          if ((/smartsurvey/.test(surveyUrl)) && (surveyUrl.indexOf('?c=') === -1)) {
```
Comment (h-lame) [inline app/assets/javascripts/surveys.js]: Doesn't the `{{currentPath}}` template param stuff handle this?  I think we can drop this.

**Comment by h-lame** [comment]
I think another set of eyes is particularly useful as I wrote the initial version of some of this so I'm not sure we're really getting 2 eyes on it when it's only you and I reviewing it.

File: spec/javascripts/surveys-spec.js (hunk 1 above)
Comment (elenatanasoiu) [inline spec/javascripts/surveys-spec.js]: Explicitly made activeWhen() to return "true"

File: spec/javascripts/surveys-spec.js [hunk 5]
```diff
@@ -90,26 +91,49 @@ describe('Surveys', function () {
       expect(GOVUK.cookie(surveys.surveySeenCookieName(defaultSurvey))).toBe('1')
     })
 
-    describe("for a 'url' survey", function () {
-      it('links to the url for a smartsurvey survey and adds the current path as a `c` param', function () {
+    describe("for a survey with a {{currentPath}} template parameter in the url", function () {
```
Comment (elenatanasoiu) [inline spec/javascripts/surveys-spec.js]: Dropped extra-explicit description

File: spec/javascripts/surveys-spec.js [hunk 6]
```diff
@@ -91,13 +91,31 @@ describe('Surveys', function () {
...
+          url: 'cleversurvey.com/default',
+          identifier: 'url-survey',
+        }
+        surveys.displaySurvey(nonSmartSurveyUrlSurvey);
+
+        expect($('#take-survey').attr('href')).toContain(nonSmartSurveyUrlSurvey.url);
+        expect($('#take-survey').attr('href')).not.toContain("?c=" + window.location.pathname);
+      })
+
+      it('records an event when showing the survey', function () {
```
Comment (elenatanasoiu) [inline spec/javascripts/surveys-spec.js]: Removed duplicate

File: app/assets/javascripts/surveys.js (hunk 3 above)
Comment (elenatanasoiu) [inline app/assets/javascripts/surveys.js]: Moved back into templateBase

File: app/assets/javascripts/surveys.js [hunk 7]
```diff
@@ -257,16 +257,15 @@
...
-      // Smart survey can record the URL of the survey link if passed through as a query param
-      if ((/smartsurvey/.test(surveyUrl)) && (surveyUrl.indexOf('?c=') === -1)) {
-        surveyUrl += '?c=' + userSurveys.currentPath()
-      }
+      var surveyUrl = userSurveys.addCurrentPathtoURL(survey)
 
       $surveyLink.attr('href', surveyUrl)
     },
 
+    addCurrentPathtoURL: function (survey) {
```
Comment (elenatanasoiu) [inline app/assets/javascripts/surveys.js]: Slimmed down to url only

File: app/assets/javascripts/surveys.js (hunk 4 above)
Comment (elenatanasoiu) [inline app/assets/javascripts/surveys.js]: Dropped in a rebase, had removed it but not pushed up.

File: doc/surveys.md (hunk 2 above)
Comment (elenatanasoiu) [inline doc/surveys.md]: Added

File: app/assets/javascripts/surveys.js [hunk 8]
```diff
@@ -58,7 +58,7 @@
   )
   var SURVEY_SEEN_TOO_MANY_TIMES_LIMIT = 2
 
-  var hmrc_survey_utm_campaign_value_map = function () {
+  var hmrcSurveyUtmCampaignValueMap = function () {
```
Comment (h-lame) [inline app/assets/javascripts/surveys.js]: oops! My ruby bad here!

File: spec/javascripts/surveys-spec.js [hunk 9]
```diff
@@ -90,26 +91,49 @@ describe('Surveys', function () {
...
         }
         surveys.displaySurvey(nonSmartSurveyUrlSurvey);
 
-        expect($('#take-survey').attr('href')).toContain(nonSmartSurveyUrlSurvey.url);
-        expect($('#take-survey').attr('href')).not.toContain("?c=" + window.location.pathname);
+        expect($('#take-survey').attr('href')).toContain("?c=" + window.location.pathname)
       })
 
+      it("allows the `currentPath` to be a relative url with a query-string", function () {
+        spyOn(surveys, 'currentPath').and.returnValue('/done/some-transaction?cachebust=1234')
```
Comment (elenatanasoiu) [inline spec/javascripts/surveys-spec.js]: Removed edge case

**Comment by elenatanasoiu** [comment]
I've run the standardJS linter over `surveys.js` and `surveys-spec.js`

File: app/assets/javascripts/surveys.js [hunk 10]
```diff
@@ -237,34 +284,19 @@
...
-      var $surveyLink = $('#take-survey')
-      var surveyUrl = survey.url
-
-      // Smart survey can record the URL of the survey link if passed through as a query param
-      if ((/smartsurvey/.test(surveyUrl)) && (surveyUrl.indexOf('?c=') === -1)) {
-        surveyUrl += '?c=' + userSurveys.currentPath()
-      }
-
-      $surveyLink.attr('href', surveyUrl)
+    addCurrentPathtoURL: function (surveyUrl) {
```
Comment (h-lame) [inline app/assets/javascripts/surveys.js]: Minor: this should be `addCurrentPathToURL`.

File: spec/javascripts/surveys-spec.js [hunk 11]
```diff
@@ -90,25 +87,109 @@ describe('Surveys', function () {
...
-      it('sets event handlers on the survey', function () {
-        spyOn(surveys, 'setURLSurveyEventHandlers')
+      it('does not inject the current path if the survey url does not contain the {{currentPath}} template parameter', function () {
         surveys.displaySurvey(urlSurvey)
-        expect(surveys.setURLSurveyEventHandlers).toHaveBeenCalledWith(urlSurvey)
+
+        expect($('#take-survey').attr('href')).toEqual(urlSurvey.url)
+      })
+
+      it('injects the url for a non-smartsurvey survey and adds current path as a `c` param', function () {
```
Comment (h-lame) [inline spec/javascripts/surveys-spec.js]: I don't think we need this test anymore, as it's a duplicate of the `'replaces the current path if the survey url contains the {{currentPath}} template parameter'` one above.  We no longer have special treatment of smartsurvey urls, so referring to non-smartsurvey urls isn't necessary.

File: spec/javascripts/surveys-spec.js [hunk 12]
```diff
@@ -90,25 +87,109 @@ describe('Surveys', function () {
       expect(GOVUK.cookie(surveys.surveySeenCookieName(defaultSurvey))).toBe('1')
     })
 
-    describe("for a 'url' survey", function () {
-      it('links to the url for a smartsurvey survey with a completion redirect query parameter', function () {
-        surveys.displaySurvey(urlSurvey)
+    describe('for a url survey', function () {
```
Comment (h-lame) [inline spec/javascripts/surveys-spec.js]: I think we should drop this describe block (but keep the tests).  We do the `{{currentPath}}` stuff for all survey types, so singling it out as something for url surveys only will give future developers the wrong idea about what we do.  Perhaps to be explicit we should duplicate the specs contained here and add them to `for an email survey` describe block as well?

File: spec/javascripts/surveys-spec.js [hunk 13]
```diff
@@ -90,25 +87,109 @@ describe('Surveys', function () {
...
       })
+
+      it('sets event handlers on the survey', function () {
+        spyOn(surveys, 'setURLSurveyEventHandlers')
+        surveys.displaySurvey(urlSurvey)
+        expect(surveys.setURLSurveyEventHandlers).toHaveBeenCalledWith(urlSurvey)
+      })
+
+      describe('without overrides for the template defaults', function () {
+        it('uses the title defined in the template data attributes', function () {
```
Comment (h-lame) [inline spec/javascripts/surveys-spec.js]: We talk about template data attributes in these tests which is a hang-over from the previous implementation where the defaults did live in data attributes.  We should reword the tests to talk about template defaults or something instead.

File: spec/javascripts/surveys-spec.js (hunk 1 above)
Comment (h-lame) [inline spec/javascripts/surveys-spec.js]: :+1:

File: doc/surveys.md (hunk 2 above)
Comment (h-lame) [inline doc/surveys.md]: :+1:

File: app/assets/javascripts/surveys.js (hunk 4 above)
Comment (h-lame) [inline app/assets/javascripts/surveys.js]: :+1:

File: app/assets/javascripts/surveys.js (hunk 3 above)
Comment (h-lame) [inline app/assets/javascripts/surveys.js]: :+1:

File: spec/javascripts/surveys-spec.js (hunk 9 above)
Comment (h-lame) [inline spec/javascripts/surveys-spec.js]: :+1:

File: spec/javascripts/surveys-spec.js (hunk 6 above)
Comment (h-lame) [inline spec/javascripts/surveys-spec.js]: :+1:

File: spec/javascripts/surveys-spec.js (hunk 1 above)
Comment (h-lame) [inline spec/javascripts/surveys-spec.js]: :+1:

File: app/assets/javascripts/surveys.js (hunk 10 above)
Comment (elenatanasoiu) [inline app/assets/javascripts/surveys.js]: Renamed

File: spec/javascripts/surveys-spec.js (hunk 12 above)
Comment (elenatanasoiu) [inline spec/javascripts/surveys-spec.js]: Split out tests for url and email for replacing the {{currentPath}} param.

File: spec/javascripts/surveys-spec.js (hunk 11 above)
Comment (elenatanasoiu) [inline spec/javascripts/surveys-spec.js]: Removed

File: spec/javascripts/surveys-spec.js (hunk 13 above)
Comment (elenatanasoiu) [inline spec/javascripts/surveys-spec.js]: Rephrased to talk about default <title|call-to-action text|success-text>

**Comment by nickcolley** [comment]
This looks good to me, I don't have full context of the code to give great opinions about it all but it seems to be internals to how Surveys work and it seems sensible. 👍

File: app/assets/javascripts/surveys.js [hunk 14]
```diff
@@ -189,14 +192,57 @@
...
+            surveySuccess: 'Thanks, we’ve sent you an email with a link to the survey.',
+            surveyFailure: 'Sorry, we’re unable to send you an email right now. Please try again later.',
+            surveyId: survey.identifier,
+            surveySource: userSurveys.currentPath(),
+            surveyUrl: userSurveys.addCurrentPathToURL(survey.url),
+          }
+          var mergedArgs = $.extend(defaultEmailArgs, survey.templateArgs)
+          return userSurveys.processTemplate(mergedArgs, EMAIL_SURVEY_TEMPLATE)
+        }
+      }
```
Comment (h-lame) [inline app/assets/javascripts/surveys.js]: This looks much nicer than the long version we had before; thanks for the suggestion @nickcolley!

File: spec/javascripts/surveys-spec.js (hunk 13 above)
Comment (h-lame) [inline spec/javascripts/surveys-spec.js]: :+1:

File: spec/javascripts/surveys-spec.js (hunk 11 above)
Comment (h-lame) [inline spec/javascripts/surveys-spec.js]: :+1:

File: spec/javascripts/surveys-spec.js (hunk 12 above)
Comment (h-lame) [inline spec/javascripts/surveys-spec.js]: :+1: - I think having these specs for each survey type is nice and explicit.  Good call.

File: app/assets/javascripts/surveys.js (hunk 10 above)
Comment (h-lame) [inline app/assets/javascripts/surveys.js]: :+1:
